### PR TITLE
Feature metric jsd (#66)

### PR DESCRIPTION
**Pull request by MooooCat**
* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Apply suggestions from code review

* jsd

* jsd

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* base更新

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Delete .idea directory

---------

## Description

The JSD single column metric.

## Motivation and Context

Since the base class of metric(Branch: feature-metric-base) has not yet been merged into the main branch, this JSD branch will first merge into the feature metric base and then merge into the main branch. 

Due to changes in some components, it is necessary to make some modifications to the examples before merging.

## Types of changes

- [ ] Maintenance (no change in code, maintain the project's CI, docs, etc.)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: sdgx/metrics/multi_table/base.py [hunk 1]
```diff
@@ -1,5 +1,4 @@
-class multiTableMetric():
-
+class multiTableMetric:
```
Comment (Wh1isper) [inline sdgx/metrics/multi_table/base.py:1]:
```suggestion
class MultiTableMetric:
```

File: sdgx/metrics/column/jsd.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+class jsd(ColumnMetric):
+    def __init__(self) -> None:
+        super().__init__()
+        self.lower_bound = 0
+        self.upper_bound = 1
+        self.metric_name = "jensen_shannon_divergence"
+
+    @classmethod
+    def calculate(cls, real_data, synthetic_data, discrete, cols):
```
Comment (Wh1isper) [inline sdgx/metrics/column/jsd.py:18]: Suggest add some type hint here

File: sdgx/metrics/column/base.py [hunk 3]
```diff
@@ -1,54 +1,52 @@
...
             synthetic_data = pd.Series(synthetic_data)
             return real_data, synthetic_data
         except:
-            raise TypeError('An error occurred while converting to pd.Series')
+            raise TypeError("An error occurred while converting to pd.Series")
         finally:
             return None, None
 
     @classmethod
     def calculate(real_data, synthetic_data):
```
Comment (Wh1isper) [inline sdgx/metrics/column/base.py:41]:
```suggestion
    def calculate(cls, real_data, synthetic_data):
```

File: sdgx/metrics/column/base.py [hunk 4]
```diff
@@ -1,54 +1,52 @@
...
         # This method should first check the input
-        # such as: 
+        # such as:
         real_data, synthetic_data = columnMetric.check_input(real_data, synthetic_data)
-        
+
         raise NotImplementedError()
 
     @classmethod
     def check_output(raw_metric_value):
```
Comment (Wh1isper) [inline sdgx/metrics/column/base.py:49]:
```suggestion
    def check_output(cls, raw_metric_value):
```

File: sdgx/metrics/column/base.py [hunk 5]
```diff
@@ -1,54 +1,52 @@
...
-    metric_name = 'Accuracy'
+    metric_name = "Accuracy"
 
     def __init__(self) -> None:
-        
         pass
-    
+
     @classmethod
     def check_input(real_data, synthetic_data):
```
Comment (Wh1isper) [inline sdgx/metrics/column/base.py:13]:
```suggestion
    def check_input(cls, real_data, synthetic_data):
```

**Review by Wh1isper — CHANGES_REQUESTED** [review]

**Comment by MooooCat** [comment]
Thanks to @Wh1isper for the carefully review. I will add some descriptive comments for input parameters in the method `calculate`.

**Comment by Wh1isper** [comment]
Actually type hint and docstring is all we need, like we did in https://github.com/hitsz-ids/synthetic-data-generator/blob/main/sdgx/data_processors/base.py


Example:

```python
# type hint example
from __future__ import annotations

@classmethod
def calculate(cls, real_data: pd.DataFrame, synthetic_data: pd.DataFrame, discrete: list[str] | None, cols: list[str] | None) -> pd.DataFrame:
"""
What is this function

Args:
  real_data (pd.DataFrame): balabala...
  ....

"""
```

**Comment by Wh1isper** [comment]
> Actually type hint and docstring are all we need
> 
> Like we did in [`main`/sdgx/data_processors/base.py](https://github.com/hitsz-ids/synthetic-data-generator/blob/main/sdgx/data_processors/base.py?rgh-link-date=2023-12-18T02%3A04%3A02Z)
> 
> Example:
> 
> ```python
> # type hint example
> from __future__ import annotations
> 
> @classmethod
> def calculate(cls, real_data: pd.DataFrame, synthetic_data: pd.DataFrame, discrete: list[str] | None, cols: list[str] | None) -> pd.DataFrame:
> """
> What is this function
> 
> Args:
>   real_data (pd.DataFrame): balabala...
>   ....
> 
> """
> ```

I'd be inclined to add this to all public methods, especially those that require subclass implementation. This is very intuitive for developers and users, and allows static checking tools to work well.

File: sdgx/metrics/column/base.py [hunk 6]
```diff
@@ -1,54 +1,52 @@
...
 
         # Check some data-types that must not be allowed
         if type(real_data) in [int, float, str]:
-            raise TypeError('real_data\'s type must not be None, int, float or str')
-        
+            raise TypeError("real_data's type must not be None, int, float or str")
+
         # if type is pd.Series, return directly
-        if type(real_data) is pd.Series:
+        if type(real_data) is pd.Series or type(real_data) is pd.DataFrame:
```
Comment (Wh1isper) [inline sdgx/metrics/column/base.py:27]:
```suggestion
        if isinstance(real_data, pd.Series ) or isinstance(real_data, pd.DataFrame):
```

File: sdgx/metrics/column/jsd.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+        self.upper_bound = 1
+        self.metric_name = "jensen_shannon_divergence"
+
+    @classmethod
+    def calculate(cls,\
+                  real_data: pd.DataFrame,\
+                  synthetic_data: pd.DataFrame,\
+                  cols: list[str] | None,\
+                  discrete: bool = True,\
+                  ) -> pd.DataFrame:
```
Comment (Wh1isper) [inline sdgx/metrics/column/jsd.py]:
```suggestion
    def calculate(
        cls,
        real_data: pd.DataFrame,
        synthetic_data: pd.DataFrame,
        cols: list[str] | None,
        discrete: bool = True,
    ) -> pd.DataFrame:
```

No need to use slash

File: sdgx/metrics/column/jsd.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+            raw_metric_value (float):  the calculated raw value of the JSD metric.
+        '''
+        instance = cls()
+        if raw_metric_value < instance.lower_bound or raw_metric_value > instance.upper_bound:
+            raise ValueError
+
+    @classmethod
+    def jensen_shannon_divergence(cls,\
+                                  p: float,\
+                                  q: float):
```
Comment (Wh1isper) [inline sdgx/metrics/column/jsd.py]:
```suggestion
    def jensen_shannon_divergence(
        cls,
        p: float,
        q: float,
    ):
```

File: sdgx/metrics/single_table/base.py [hunk 9]
```diff
@@ -8,21 +7,19 @@ class singleTableMetric():
...
 
-    # not a class method 
-    def calculate(real_data, synthetic_data):
-
+    # not a class method
+    def calculate(self, real_data, synthetic_data):
         raise NotImplementedError()
 
     @classmethod
     def check_output(raw_metric_value):
```
Comment (Wh1isper) [inline sdgx/metrics/single_table/base.py]:
```suggestion
    def check_output(cls, raw_metric_value):
```

**Review by Wh1isper — CHANGES_REQUESTED** [review]

File: sdgx/metrics/column/jsd.py (hunk 7 above)
Comment (MooooCat) [reply sdgx/metrics/column/jsd.py]: Noticed, parameters in multiple line will not use slash.

**Review by Wh1isper — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**